interfaces/hotplug/udevadm.go

```diff
@@ -0,0 +1,124 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package hotplug
+
+import (
+	"bufio"
+	"fmt"
+	"os/exec"
+	"strings"
+)
+
+var udevadmBin = `udevadm`
+
+// RunUdevadm enumerates all devices by parsing 'udevadm info -e' command output and reports them
+// via devices channel. The devices channel gets closed to indicate that all devices were processed.
+// Parsing errors are reported to parseErrors channel. Fatal errors encountered when starting udevadm
+// are reported by the error return value.
+func RunUdevadm(devices chan<- *HotplugDeviceInfo, parseErrors chan<- error) error {
+	cmd := exec.Command(udevadmBin, "info", "-e")
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		close(devices)
+		return err
+	}
+
+	rd := bufio.NewScanner(stdout)
+	if err := cmd.Start(); err != nil {
+		close(devices)
+		return err
+	}
+
+	go func() {
+		defer close(devices)
+
+		env := make(map[string]string)
+
+		var deviceBlock bool
+		for rd.Scan() {
+			// udevadm export output is divided in per-device blocks, blocks are separated
+			// with empty lines, each block starts with device path (P:) line, within a block
+			// there is one attribute per line, example:
+			//
+			// P: /devices/virtual/workqueue/nvme-wq
+			// E: DEVPATH=/devices/virtual/workqueue/nvme-wq
+			// E: SUBSYSTEM=workqueue
+			// <empty-line>
+			// P: /devices/virtual/block/dm-1
+			// N: dm-1
+			// S: disk/by-id/dm-name-linux-root
+			// E: DEVNAME=/dev/dm-1
+			// E: USEC_INITIALIZED=8899394
+			// <empty-line>
+			line := rd.Text()
+			if line == "" {
+				deviceBlock = false
+				if len(env) > 0 {
+					outputDevice(env, devices, parseErrors)
+					env = make(map[string]string)
+				}
+				continue
+			}
+			if strings.HasPrefix(line, "P: ") {
+				deviceBlock = true
+				continue
+			}
+
+			// any property we find needs to belong to a device "block" started by "P: "
+			if deviceBlock == false {
+				parseErrors <- fmt.Errorf("no device block marker found before %q", line)
+				continue
+			}
+
+			// We are only interested in 'E' properties as they carry all the interesting data,
+			// including DEVPATH and DEVNAME which seem to mirror the 'P' and 'N' values.
+			if strings.HasPrefix(line, "E:") {
+				if kv := strings.SplitN(line[3:], "=", 2); len(kv) == 2 {
+					env[kv[0]] = kv[1]
+				} else {
+					parseErrors <- fmt.Errorf("failed to parse udevadm output %q", line)
+				}
+			}
+		}
+
+		// eof, flush remaining device
+		if len(env) > 0 {
```

Comment:
You may end up with incomplete device (eg. with only devpath or devname) here if an error when reading the string occurred. I'd say that the error check and this if need to swap places.

Reply:
Good point, thanks. Re-ordered and made it nullify the env in such case.